interface/src/app/gene/enriched_signatures.js

```diff
@@ -0,0 +1,222 @@
+          }
+        }
+        matchedGenesBySignature.push(matchedGenes);
+
+        // Four parameters are needed in the formula:
+        //     hyperGeometricTest(k, m, n, N)
+        // k: the number of genes related to sigID in genesInUrl;
+        // m: the number of genes in genesInUrl;
+        // n: the total number of genes related to sigID;
+        // N: the total number of genes.
+        pValue = 1 - MathFuncts.hyperGeometricTest(k, m, n, N);
+        pValueArray.push(pValue);
+      });
+
+      var significantSignatures = [];
+      var correctedPValues = MathFuncts.multTest.fdr(pValueArray);
+      correctedPValues.forEach(function(element, index) {
+        var correctedPValue, sigID;
+        if (element < self.pValueCutoff) {
+          correctedPValue = element.toPrecision(pValueDigits);
+          sigID = signatureIDs[index];
+          significantSignatures.push({
+            'url': '#/signature/' + sigID,
+            'name': self.signatures[sigID],
+            'participationType': typeName,
+            'genes': matchedGenesBySignature[index].join(', '),
+            'pValue': correctedPValue
+          });
+        }
+      });
+
+      significantSignatures.sort(function(a, b) {
+        return a.pValue - b.pValue;
+      });
+
+      return significantSignatures;
+    };
+
+    // groupedGenes is an object that behaves like a 2-D table of genes.
+    // groupedGenes[participationTypeID][signatureID] is an object whose
+    // keys are gene IDs and values are gene names of those genes that
+    // participate with signatureID and with participationTypeID.
+    var groupedGenes = Object.create(null);
+
+    // Main function to calculate the enrichment.
+    var getEnrichedSignatures = function() {
+      participations.forEach(function(element) {
+        var sigID = element.signature;
+        if (!self.signatures[sigID]) {
+          return;
+        } // ignore signatures that are not in current mlmodel
+
+        var typeName = element.participation_type.name;
+        if (!groupedGenes[typeName]) {
+          groupedGenes[typeName] = Object.create(null);
+        }
+
+        if (!groupedGenes[typeName][sigID]) {
+          groupedGenes[typeName][sigID] = Object.create(null);
+        }
+        geneID = element.gene.id;
+        geneName = (element.gene.standard_name ?
+                    element.gene.standard_name :
+                    element.gene.systematic_name);
+        groupedGenes[typeName][sigID][geneID] = geneName;
+      });
+
+      var participationTypes = Object.keys(groupedGenes);
+      participationTypes.forEach(function(typeName) {
+        var genesBySignatures = groupedGenes[typeName];
+        var enrichment = calculateEnrichments(typeName, genesBySignatures);
+        for (var i = 0; i < enrichment.length; i++) {
+          self.enrichedSignatures.push(enrichment[i]);
+        }
+      });
+      self.statusMessage = '';
+      self.inProgress = false;
```

Comment:
I like the idea of using `finally`. Actually I realized that the current error handling has a bug. Since each promise has its own error handling function, even if any of these promises is rejected, `getEnrichedSignatures` function will still be called. So I need to throw an exception inside each error handling function so that the error can be detected by `catch()` and `getEnrichedSignatures` will be skipped.

Reply:
Even better. That looks really great now! Much easier to follow what is going on, and less repeated code. Even the curmudgeonly Codeclimate seems happier about it.